### PR TITLE
[vcpkg] Introduce new vcpkg option VCPKG_WARNING_FLAGS / VCPKG_OPTIMIZED_FLAGS_DEBUG / VCPKG_OPTIMIZED_FLAGS_RELEASE

### LINKED ISSUE #26699: [Suggestion]: Remove `/W3` being used when installing packages on Windows
Hi,

This might be the wrong place to post this, so here goes.

I have noticed that when `vcpkg` installs a package on Windows, it adds `/W3` to the CXXFLAGS, see below. https://github.com/microsoft/vcpkg/blob/927006b24c3a28dfd8aa0ec5f8ce43098480a7f1/scripts/toolchains/windows.cmake#L54 
This causes the following warning to be generated by MSVC 2019 when we in our projects use `/W4`:

```
Command line warning D9025: overriding '/W3' with '/W4'
```

This warning will be generated for each translation unit. When looking through the logs for troubleshooting a failed installation, this warning produces noise.

Why does `vcpkg` set this warning at all? Should it not be up to the individual projects to decide their warning level preference?

**Proposal**
vcpkg should not set warning flags for any toolchain and leave that responsibility up to each package maintainer.

EDIT: Added link to line in question.

### PR DESCRIPTION
**Pull request by JackBoosY**
When users use vcpkg as a dependency manager to build custom projects, it is unfortunate that vcpkg sets the default compiler warning level and compiler optimization level which does NOT match the user's personal settings.

We should provide some new vcpkg macros to override these compiler flags.

Fixes https://github.com/microsoft/vcpkg/issues/26699

### CODE REVIEW

File: docs/users/triplets.md [hunk 1]
```diff
@@ -84,6 +84,21 @@ This option also has forms for configuration-specific and C flags:
 - `VCPKG_C_FLAGS_DEBUG`
 - `VCPKG_C_FLAGS_RELEASE`
 
+### VCPKG_WARNING_FLAGS
+Sets compiler warning level to be used while building C and CXX files.
+
+The default value is `/W3` on Windows.
```
Comment (JackBoosY) [inline docs/users/triplets.md:90]: @Neumann-A Should we also set the default values in other toolchain files?

**Comment by JackBoosY** [comment]
The compiler flags stream:
```
Toolchains -> CMAKE_*_FLAGS -> vcpkg_cmake_configure / vcpkg_configure_cmake
                            -> z_vcpkg_get_cmake_vars -> vcpkg_configure_make / vcpkg_configure_gn / vcpkg_configure_meson / vcpkg_configure_qmake etc.
```

**Comment by Neumann-A** [comment]
Simply no. Do not add more switches..... the correct mechanism is to provide your own toolchain if you don't like what vcpkg defaults to. 

Sorry but that is just something people have to learn. Maybe vcpkg just needs to provide better examples/documentation how to achieve that.

**Comment by JackBoosY** [comment]
> Simply no. Do not add more switches..... the correct mechanism is to provide your own toolchain if you don't like what vcpkg defaults to.

At least vcpkg should provide a base toolchain or configuration.
And that's why vcpkg provides `VCPKG_*_FLAGS`.

**Comment by Neumann-A** [comment]
> should provide a base toolchain 

it does and using it for everything is an error as should have been apparent when the uwp toolchain was split out of the normal windows one. 

> And that's why vcpkg provides VCPKG_*_FLAGS.

maybe vcpkg should never have provided these variables. But what you want to achieve here can already be achieved by setting `VCPKG_*_FLAGS` even though this will through a warning which probably could be activly ignored. 

Adding more of those flags will just add more unnecessary clutter to the triplet, it gives more stuff to learn and just delays the inevitable necessity to learn how to pass a custom toolchain to vcpkg. 

Furthermore you current implementation does not correctly take care of forwarding the variables to `try_compile`. 
Secondly you didn't touch `vcpkg_configure_cmake/cmake_configure` to actually forward the variables from the triplet to cmake which effectively means they are not used.
Thirdly, this just adds code everywhere which needs to be maintained with little to no benefit.

So the answer is: If you want to modify what vcpkg provides in more detail -> provide a custom toolchain with the required domain knowledge.

**Comment by Osyotr** [comment]
But what is the point in setting warning level when building packages?

**Comment by JackBoosY** [comment]
> So the answer is: If you want to modify what vcpkg provides in more detail -> provide a custom toolchain with the required domain knowledge.

This is obviously not user-friendly, and the time consumption of learning to use vcpkg macros is always shorter than learning to write a toolchain: users may not care what the toolchain does but care about the flags.

I agree that these flags need to be separated according to the OS and not exist in each toolchains.

**Comment by Neumann-A** [comment]
> always shorter than learning

Thanks for confirming that users are lazy. 

>  and the time consumption of learning to use vcpkg macros

teaching custom triplet/toolchain combo in one go is two flies with one stone. You have to teach triplets customization any way to use these flags. 

> This is obviously not user-friendly,

Sometimes being not overlay user-friendly can be more user-friendly in the long run. Adding flags is just one step but it typically goes into the direction of requiring more customization in the end which is why `VCPKG_*_FLAGS` is more than enough.

>  users may not care what the toolchain does

They should care since it is an essential part of using CMake/vcpkg.

**Comment by JackBoosY** [comment]
Let's leave the decision to the maintainers.

**Comment by dg0yt** [comment]
> But what is the point in setting warning level when building packages?

i.e. shoud `/w...` be removed?

**Comment by Osyotr** [comment]
> i.e. shoud `/w...` be removed?

Yeah. Does someone read build log if the build succeeds?

**Comment by JackBoosY** [comment]
cc @BillyONeal @ras0219-msft

**Comment by JackBoosY** [comment]
@FrankXie05 baseline regressions:
kf5solid:x64-windows-static-md:
```
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_subscribe referenced in function idevice_event_subscribe
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_unsubscribe referenced in function idevice_event_unsubscribe
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_device_list referenced in function idevice_get_device_list
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_device_list_free referenced in function idevice_get_device_list
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_device referenced in function idevice_new_with_options
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_connect referenced in function idevice_connect
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_disconnect referenced in function idevice_disconnect
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_send referenced in function internal_connection_send
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_recv_timeout referenced in function internal_connection_receive_timeout
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_recv referenced in function internal_connection_receive
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_set_socket_type referenced in function idevice_set_socket_type
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_socket_type referenced in function idevice_get_socket_type
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_set_tcp_endpoint referenced in function idevice_set_tcp_endpoint
libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_tcp_endpoint referenced in function idevice_get_tcp_endpoint
libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_read_buid referenced in function userpref_read_system_buid
libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_read_pair_record referenced in function userpref_read_pair_record
libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_save_pair_record_with_device_id referenced in function userpref_save_pair_record
libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_delete_pair_record referenced in function userpref_delete_pair_record
bin\solid-hardware5.exe : fatal error LNK1120: 18 unresolved externals
```
lack of usbmuxd library I think?

osg:x64-windows-static-md:
```
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(323): error C2061: syntax error: identifier 'GLint64'
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(324): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(324): error C2143: syntax error: missing ',' before '*'
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(325): error C2061: syntax error: identifier 'GLuint64'
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(326): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
```
I saw this regressions before, please search.

**Comment by dg0yt** [comment]
> osg:x64-windows-static-md:

#28087

**Comment by FrankXie05** [comment]
@JackBoosY About regressions `kf5solid:x64-windows-static-md`, we are investigating, can you help to fix it? :)

**Comment by Neumann-A** [comment]
> About regressions kf5solid:x64-windows-static-md, we are investigating, can you help to fix it? :)

@FrankXie05: You are really still investigating this one? read #28022 ... it is an optional dependency which kf5solid picks up depending on CI build order since it is not controlled currently.

**Comment by LilyWangLL** [comment]
> @FrankXie05 baseline regressions: kf5solid:x64-windows-static-md:
> 
> ```
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_subscribe referenced in function idevice_event_subscribe
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_unsubscribe referenced in function idevice_event_unsubscribe
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_device_list referenced in function idevice_get_device_list
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_device_list_free referenced in function idevice_get_device_list
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_device referenced in function idevice_new_with_options
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_connect referenced in function idevice_connect
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_disconnect referenced in function idevice_disconnect
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_send referenced in function internal_connection_send
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_recv_timeout referenced in function internal_connection_receive_timeout
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_recv referenced in function internal_connection_receive
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_set_socket_type referenced in function idevice_set_socket_type
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_socket_type referenced in function idevice_get_socket_type
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_set_tcp_endpoint referenced in function idevice_set_tcp_endpoint
> libimobiledevice-1.0.lib(idevice.c.obj) : error LNK2019: unresolved external symbol usbmuxd_get_tcp_endpoint referenced in function idevice_get_tcp_endpoint
> libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_read_buid referenced in function userpref_read_system_buid
> libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_read_pair_record referenced in function userpref_read_pair_record
> libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_save_pair_record_with_device_id referenced in function userpref_save_pair_record
> libimobiledevice-1.0.lib(userpref.c.obj) : error LNK2019: unresolved external symbol usbmuxd_delete_pair_record referenced in function userpref_delete_pair_record
> bin\solid-hardware5.exe : fatal error LNK1120: 18 unresolved externals
> ```
> 
> lack of usbmuxd library I think?
> 
> osg:x64-windows-static-md:
> 
> ```
> D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(323): error C2061: syntax error: identifier 'GLint64'
> D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(324): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
> D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(324): error C2143: syntax error: missing ',' before '*'
> D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(325): error C2061: syntax error: identifier 'GLuint64'
> D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(326): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
> ```
> 
> I saw this regressions before, please search.

`kf5solid:x64-windows-static-md`: #28022

**Comment by FrankXie05** [comment]
@JackBoosY Please resolve file conflicts. 

Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". 
          That way, I can be aware that you've responded since you can't modify the tags.

**Pull request closed without merging**